File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        vector<int> nge(n,-1);
        int maxx=prices[n-1];
        for(int i=n-2;i>=0;i--){
            nge[i]=maxx;
            maxx=max(maxx,prices[i]);
        }

        int ans=0;
        int idx=0;
        for(int i=0;i<n;i++){
            if(nge[i]!=-1 && (nge[i]-prices[i])>ans){
               ans=nge[i]-prices[i];
               idx=i;
               cout<<i<<" "<<ans<<endl;
            }
        }
        return ans;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        
        int n=prices.length;
        if(n==0 || n==1) return 0;

        int[] suf=new int[n];
        suf[n-1]=prices[n-1];
        int maxx=suf[n-1];
        for(int i=n-2;i>=0;i--){
            suf[i]=maxx;
            if(prices[i]>maxx) maxx=prices[i];
        }
        int ans=0;
        for(int i=0;i<n;i++){
            ans=Math.max(ans,suf[i]-prices[i]);
        }
        return ans;
    }
}